### ISSUE TITLE
a11y][Violation][1.4.3 Contrast (Minimum)] Accessibility issues reported on JupyterLab UI using IBM Equal Access Accessibility Checker tool

### ISSUE THREAD

**Issue by nkn2022** [jupyterlab/jupyterlab#16888]
## Description

I am running an accessibility test suite with Jupyter using the **IBM Equal Access Accessibility Checker** tool on Chrome browser. The results look good overall, only a few things are missing to reach a first formal level of compliance. So kindly help look into these. 

Issues reported were violating **WCAG 2.2** requirements. 
Ref: https://www.ibm.com/able/requirements/checker-rule-sets

Listing of violations categorized based on the checkpoints. 
Full report (includes information on element locations where issue were found): 
[Accessibility_Report-Untitled.ipynb - JupyterLab.xlsx](https://github.com/user-attachments/files/17383160/Accessibility_Report-Untitled.ipynb.-.JupyterLab.xlsx)

# Category: `1.4.3 Contrast (Minimum)`
Ref: https://www.ibm.com/able/requirements/requirements/?version=v7_3#1_4_3

****Violation of Rule:** `text_contrast_sufficient`**

> List of issues with elements that violate this rule with a screenshots
> > **Issue 1** 
> > Text contrast of 4.18 with its background is less than the WCAG AA minimum requirements for text of size 12px and weight of 400
> > **Element location:** 
> > `<jp-button minimal="" appearance="stealth" current-value="" tabindex="0" title="New Launcher (⇧ ⌘ L)" data-command="launcher:create" aria-label="New Launcher (⇧ ⌘ L)" aria-disabled="false" class="jp-ToolbarButtonComponent">`
> > **Screenshot:**
> > <img width="468" alt="image" src="https://github.com/user-attachments/assets/ec793a00-f59b-48e9-9b9d-184f5194bb15">

> > **Issue 2** 
> > Text contrast of 4.26 with its background is less than the WCAG AA minimum requirements for text of size 13px and weight of 400
> > **Element location:** 
> > `<span class="ͼu">`
> > **Screenshot:**
> > <img width="468" alt="image" src="https://github.com/user-attachments/assets/561b8e5a-b862-4afb-970e-adf965ebfca6">

## Reproduce

- Install Chrome extension for IBM accessibility checker tool. Ref: https://www.ibm.com/able/toolkit/verify/automated
- Open JupyterLab UI on Chrome browser. Open `notebook` file to edit.
- Run the tool and we can see the above `Violation`s reported.

Here is the screenshot of the UI scanned for the attached report:

<img width="1534" alt="image" src="https://github.com/user-attachments/assets/0e0cbc5c-cf21-4393-97fe-b566da24b58b">

## Expected behavior

No violations are reported on the scanning to reach a first formal level of compliance.

## Context

- Operating System and version: MacOS Sonoma Version 14.5
- Browser and version: Chrome Version 129+
- JupyterLab version: Version 4.2.5

**Comment by JasonWeill** [comment]
Might be good to check the same elements using JupyterLab's high contrast theme, which is specifically intended for users who need higher contrast than the default theme offers. Do you still see the problems in that case?

Also, we might want to update the [JupyterLab accessibility statement](https://jupyterlab.readthedocs.io/en/latest/getting_started/accessibility.html) to mention contrast, because we've had questions like this come up before.

**Comment by krassowski** [comment]
https://webaim.org/resources/contrastchecker/

![image](https://github.com/user-attachments/assets/54fdf6c0-ffcc-4885-ab2d-477a537934ae)

**Comment by JasonWeill** [comment]
Code coloring is already covered by #14255.

Link styling is in  #14166.

Adding tests for a11y as part of PR timelines is covered in #9742, which also mentions IBM's accessibility checker.

**Comment by JasonWeill** [comment]
Retitled this to cover only sub-issue 1, about the "New" (+) button in the file browser.

**Comment by nkn2022** [comment]
@krassowski I think you had the foreground and backgroud switched. Can you double check?
Also I see that the button is using hex `#1976D2` for the color not `#0078CE`. 

> <img width="409" alt="image" src="https://github.com/user-attachments/assets/36501cec-4ebe-4c13-8dfd-3f018c524c6e">


when I check with that I get a different value `4:6:1`. 

> <img width="697" alt="image" src="https://github.com/user-attachments/assets/bde1a18f-c091-4a8b-9a3a-4169caa0c323">

**Comment by krassowski** [comment]
I used a color picker which might have been slightly off but my point is that **4.18** in your report does not agree with what I see:

> Text contrast of **4.18** with its background is less than the WCAG AA minimum requirements for text of size 12px and weight of 400

Since 4.6 is greater than 4.5 which is the WCAG AA minimum requirements for text of size 12px and weight of 400.

**Comment by krassowski** [comment]
> I think you had the foreground and backgroud switched. Can you double check?

This does not matter because the contrast calculations win WCAG 2.x are not very clever and do not account for the difference between background and foreground at all (which bites Jupyter which uses a beautiful orange background but some folks came and changed perfectly readable white text to unreadable black to satisfy arbitrary rules of WCAG 2 contrast checkers, see https://github.com/jupyter/jupyter.github.io/issues/587 - this is only getting fixed in WCAG 3 with APCA)

![image](https://github.com/user-attachments/assets/e7f24e87-5c23-460d-8351-44786a9e119e)

**Comment by krassowski** [comment]
Using another color picker I again arrive at `#0078CE` (which is why I think you cannot relay on automated checkers that match as they will look say at the `background` attribute but ignore `opacity` or CSS filters which modify the color), at least on https://jupyter.org/try-jupyter/lab/

![image](https://github.com/user-attachments/assets/07b95b35-4c3f-47c9-83bc-f27bb0f6579c)

**Comment by nkn2022** [comment]
Thanks. Yes, that may be the reason here. I will confirm soon.

**Comment by krassowski** [comment]
I see what happened with the contrast checker that you used. It indeed blindly extracted the `background` and `color` even though there is no text, but only white SVG. Understandably it gets lost with web components/svg and just attempts to use what it sees in the CSS:

![image](https://github.com/user-attachments/assets/b89443b1-dbc1-4349-9b54-64037dbc9750)

![image](https://github.com/user-attachments/assets/64c255ac-ed4a-4d06-80fe-8a1aeb52b524)

Since 4.19 is very close to 4.18 that you reported I think this is what happened. I think this is not a real issue then.

**Comment by nkn2022** [comment]
I agree with you. I will close this for now. Thank you for the investigation.

**Comment by nkn2022** [comment]
Adding the response from the tools team here for reference. 

> Phill Jenkins
> :beequala11y:  [16 minutes ago](https://ibm-analytics.slack.com/archives/C036P1CTN/p1730306840335889?thread_ts=1730297552.305909&cid=C036P1CTN)
> I'll might open an issue with the Checker to compare how it calculates the contrast vs how other tools do it, such as the [CCA tool](https://www.tpgi.com/color-contrast-checker/).point is sufficient, while the edges are not.  So, in my opinion the Checker is correct in flagging this as a contrast violation.
> ![image](https://github.com/user-attachments/assets/31de6fa7-e2eb-4d2a-9879-8ae5f94d6fc7)
> 
> NISHA NAIR
>   12 minutes ago
> Okay. Thank you very much for checking. Should we chase the team to fix this gradation problem then? Not sure how critical this is from a user ’s perspective. (edited) 
> 
> 
> NISHA NAIR
>   [10 minutes ago](https://ibm-analytics.slack.com/archives/C036P1CTN/p1730307143016099?thread_ts=1730297552.305909&cid=C036P1CTN)
> I will put your comment in the external issue so they are aware of this issue.
> 
> 
> Phill Jenkins
> :beequala11y:  [3 minutes ago](https://ibm-analytics.slack.com/archives/C036P1CTN/p1730307583066009?thread_ts=1730297552.305909&cid=C036P1CTN)
> Yes I would recommend that the SVG icon (it doesn't look like real text "+") get improved. At 100% zoom it fails, so the low-vision may miss that there is a "+" over that blue button.  But since there is only one "blue button", seeing the "+" is not a critical.
> When zooming out the SVG it looks a lot better, but it doesn't need to have that gradation in my opinion.
> :+1:
> 1

**Comment by tombrunet** [comment]
The tool does in fact handle opacity, but that's not what's going on here. The reason this is flagging is that the foreground color of the button is specified as a transparent black.
<img width="649" alt="image" src="https://github.com/user-attachments/assets/656ab5df-8370-4cb1-bccb-38620557459d">
The SVG is overriding this color (which is why it's flagging the button and not the svg). We'll have to look to see if we can handle this better with the SVG defined colors, but if you have any buttons with the same CSS and there's text instead of SVG, it would have an issue.

**Comment by krassowski** [comment]
Thanks @tombrunet, this confirms my guess from https://github.com/jupyterlab/jupyterlab/issues/16888#issuecomment-2447314562 then. Indeed it could be an issue, but it's only this single button which has blue background (the rule targets it via command id) so it is guaranteed to have a white icon rather than any text.
